libtiledbsoma/src/reindexer/reindexer.h

```diff
@@ -0,0 +1,85 @@
+/**
+ * @file   reindexer.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the a pybind11 api into SOMA C++ library.
+ */
+
+#pragma once
+#include <assert.h>
+#include <thread_pool/thread_pool.h>
+#include <unistd.h>
+#include "khash.h"
+
+namespace tiledbsoma {
+
+// Typedef for a 64-bit khash table
+KHASH_MAP_INIT_INT64(m64, int64_t)
+
+class IntIndexer {
+   public:
+    /**
+     * Perform intitalization of hash and threadpool
+     * @param keys pointer to key array of 64bit integers
+     * @param size yhr number of keys in the put
+     * @param threads number of threads in the thread pool
+     */
+    void map_locations(const int64_t* keys, int size, int threads = 4);
+    /**
+     * @return the number of keys
+     */
+    int size();
+    /**
+     * Used for parallel lookup using khash
+     * @param keys array of keys to lookup
+     * @param result array for lookup results
+     * @param size // Number of key array
+     * @return and array of looked up value (same size as keys)
+     */
+    void lookup(int64_t* keys, int64_t* results, int size);
+    IntIndexer() {};
+    /**
+     * Constructor with the same arguments as map_locations
+     */
+    IntIndexer(const int64_t* keys, int size, int threads);
+    virtual ~IntIndexer();
+
+   private:
+    int thread_count_;
+    int size_;
+    /*
+     * The created 64bit hash table
+     */
+    khash_t(m64) * hash_;
+    /*
+     * TileDB threadpool
+     */
+    std::shared_ptr<tiledbsoma::ThreadPool> tiledb_thread_pool_ = nullptr;
```

Comment:
We could probably avoid needing to install the `thread_pool.h` and `khash.h` headers altogether by forward declaring `ThreadPool` and `khash_t(m64)` since these are both pointers and then move the includes into the cc. This way we can keep the threadpool and khash code private.